python/lsst/obs/base/filters.py

```diff
@@ -0,0 +1,129 @@
+
+        WARNING: Resets the afw filter list before defining the filters.
+        """
+        lsst.afw.image.utils.resetFilters()
+        for filter in sorted(self._filters):
+            filter.defineFilter()
+
+
+@dataclasses.dataclass(frozen=True, order=True)
+class FilterDefinition:
+    """The definition of an instrument's filter bandpass.
+
+    A temporary stand-in for the afw.image Filter classes used to
+    share initialization between Gen2 Mappers and Gen3 Instruments.
+    """
+    name: str
+    """The name of the filter; used to define the unique `afw.image.Filter`.
+
+    WHAT IS THIS ACTUALLY!!!?!?!?!?!!?!
+
+    For HSC, it is neither physical nor abstract: often it is the same as
+    the abstract filter, sometimes it is a shortened version of the physical
+    filter but not the same as the abstract filter.
+    """
+
+    physical_filter: str
+    """The name of a filter associated with a particular instrument: unique for
+    each piece of glass. This should match the exact filter name used in the
+    observatory's metadata.
+    This name is added to the aliases.
+    """
+
+    lambdaEff: float
+    """The effective wavelength of this filter (nm)"""
+
+    abstract_filter: str = None
+    """The name of a filter not associated with a particular instrument
+    (e.g. `r` for the SDSS Gunn r-band, which could be on SDSS, LSST, or HSC).
+    If there is no generic filter name for this, use None.
+    If this does not match `name`, it is added to the aliases.
+    """
+
+    lambdaMin: float = np.nan
+    """The minimum wavelength of this filter (nm; defined as 1% throughput)"""
+    lambdaMax: float = np.nan
+    """The maximum wavelength of this filter (nm; defined as 1% throughput)"""
+
+    alias: set = frozenset()
+    """Alternate names for this filter."""
+
+    def defineFilter(self):
+        """Declare the filters via afw.image.Filter.
+        """
+        afw_aliases = set(self.alias)
+        # always add physical_filter to the aliases
+        afw_aliases.add(self.physical_filter)
+        # Only add abstract_filter if it is different from the generic name.
+        if self.abstract_filter is not None and self.name != self.abstract_filter:
+            afw_aliases.add(self.abstract_filter)
+        lsst.afw.image.utils.defineFilter(self.name,
+                                          lambdaEff=self.lambdaEff,
+                                          lambdaMin=self.lambdaMin,
+                                          lambdaMax=self.lambdaMax,
+                                          alias=afw_aliases)
+
+    def __hash__(self):
+        return hash(self.name)
+
+    def __str__(self):
+        return self.name
+
+    def __repr__(self):
+        rangestr = ""
+        if not np.isnan(self.lambdaMin) and not np.isnan(self.lambdaMax):
+            rangestr = f" (self.lambdaMin, self.lambdaMax) "
+        return f"{self.name}, {self.lambdaEff}{rangestr}"
```

Comment:
Shouldn't repr return something that looks vaguely like an object constructor string?

Reply:
I forgot that `dataclass` gives str and repr: I'll just use its `repr`, but I wonder if we want a simpler `str` that doesn't include everything (or doesn't include the default-valued things?).

I overrode `hash` so that when you create a `FilterDefinition`, you can use `alias={'blah'}`, not `alias=frozenset('blah')`, because the `FilterDefinitions` needs to hash it and `set` is not hashable. I want `FilterDefinition` to be frozen/immutable, but it would be annoying to have to use `frozenset` when defining the aliases.

Reply:
I sorted out a solution for `hash` (`__post_init__` casting of `alias` to a `frozenset`) and will just use the default `str` until we decide we want something simpler/smarter.